File: INTeract-ive-Agent/typescript/src/tools/refactor-tools.ts
/**
 * Refactoring analysis tools for the agent
 */

import { tool, createSdkMcpServer } from '@anthropic-ai/claude-agent-sdk';
import { z } from 'zod';
import * as fs from 'fs/promises';
import * as path from 'path';
import ts from 'typescript';

/**
 * Find duplicate code blocks across files
 */
export async function _findDuplicatesImpl(directory: string, minLines: number = 5, filePattern?: string) {
    // Simple duplicate detection using line hashing
    // Simple duplicate detection using line hashing
    const pattern = filePattern || '*';
    const files: string[] = [];

    async function walkDir(dir: string): Promise<void> {
      const entries = await fs.readdir(dir, { withFileTypes: true });
      for (const entry of entries) {
        const fullPath = path.join(dir, entry.name);
        if (entry.isDirectory() && !entry.name.startsWith('.') && entry.name !== 'node_modules') {
          await walkDir(fullPath);
        } else if (entry.isFile() && (pattern === '*' || entry.name.endsWith(pattern.replace('*', '')))) {
          files.push(fullPath);
        }
      }
    }

    await walkDir(directory);

    const blockHashes = new Map<string, Array<{ file: string; startLine: number; lines: string[] }>>();

    for (const file of files) {
      try {
        const content = await fs.readFile(file, 'utf-8');
        const lines = content.split('\n');

        for (let i = 0; i <= lines.length - minLines; i++) {
          const block = lines.slice(i, i + minLines);
          const blockText = block.join('\n').trim();

          if (blockText.length < 50) continue; // Skip tiny blocks

          const hash = Buffer.from(blockText).toString('base64').substring(0, 32);

          if (!blockHashes.has(hash)) {
            blockHashes.set(hash, []);
          }
          blockHashes.get(hash)!.push({
            file: path.relative(directory, file),
            startLine: i + 1,
            lines: block,
          });
        }
      } catch {
        // Skip files that can't be read
      }
    }

    const duplicates = Array.from(blockHashes.entries())
      .filter(([, locations]) => locations.length > 1)
      .map(([, locations]) => ({
        occurrences: locations.length,
        locations: locations.map(l => ({
          file: l.file,
          startLine: l.startLine,
          preview: l.lines.slice(0, 3).join('\n'),
        })),
      }))
      .slice(0, 20); // Limit results

    return {
      filesScanned: files.length,
      duplicatesFound: duplicates.length,
      duplicates,
    };
  }

const findDuplicates = tool(
  'find_duplicates',
  'Find duplicate or similar code blocks across files in a directory',
  {
    directory: z.string().describe('Directory to search for duplicates'),
    minLines: z.number().optional().describe('Minimum lines to consider as duplicate (default: 5)'),
    filePattern: z.string().optional().describe('Glob pattern for files to check (e.g., "*.ts")'),
  },
  async ({ directory, minLines = 5, filePattern }) => {
    const result = await _findDuplicatesImpl(directory, minLines, filePattern);
    return {
      content: [{
        type: 'text' as const,
        text: JSON.stringify(result, null, 2),
      }],
    };
  }
);

/**
 * Analyze code complexity
 */
export async function _analyzeComplexityImpl(filePath: string) {
    const content = await fs.readFile(filePath, 'utf-8');
    const sourceFile = ts.createSourceFile(filePath, content, ts.ScriptTarget.Latest, true);

    let cyclomaticComplexity = 1;
    let functionCount = 0;
    const longFunctions: Array<{ name: string; lines: number }> = [];

    function visit(node: ts.Node) {
        // Cyclomatic Complexity increments
        switch (node.kind) {
            case ts.SyntaxKind.IfStatement:
            case ts.SyntaxKind.WhileStatement:
            case ts.SyntaxKind.DoStatement:
            case ts.SyntaxKind.ForStatement:
            case ts.SyntaxKind.ForInStatement:
            case ts.SyntaxKind.ForOfStatement:
            case ts.SyntaxKind.ConditionalExpression:
            case ts.SyntaxKind.CaseClause:
            case ts.SyntaxKind.CatchClause:
                cyclomaticComplexity++;
                break;
            case ts.SyntaxKind.BinaryExpression:
                const binaryExpr = node as ts.BinaryExpression;
                if (
                    binaryExpr.operatorToken.kind === ts.SyntaxKind.AmpersandAmpersandToken ||
                    binaryExpr.operatorToken.kind === ts.SyntaxKind.BarBarToken ||
                    binaryExpr.operatorToken.kind === ts.SyntaxKind.QuestionQuestionToken
                ) {
                    cyclomaticComplexity++;
                }
                break;
        }

        // Function detection and length
        if (
            ts.isFunctionDeclaration(node) ||
            ts.isMethodDeclaration(node) ||
            ts.isArrowFunction(node) ||
            ts.isFunctionExpression(node)
        ) {
            functionCount++;
            const { line: startLine } = sourceFile.getLineAndCharacterOfPosition(node.getStart());
            const { line: endLine } = sourceFile.getLineAndCharacterOfPosition(node.getEnd());
            const lines = endLine - startLine + 1;

            if (lines > 50) {
                let name = 'anonymous';
                if (ts.isFunctionDeclaration(node) && node.name) name = node.name.text;
                if (ts.isMethodDeclaration(node) && ts.isIdentifier(node.name)) name = node.name.text;
                if (ts.isVariableDeclaration(node.parent) && ts.isIdentifier(node.parent.name)) name = node.parent.name.text;
                
                longFunctions.push({ name, lines });
            }
        }

        ts.forEachChild(node, visit);
    }

    visit(sourceFile);

    const lines = content.split('\n');
    return {
        file: filePath,
        totalLines: lines.length,
        codeLines: lines.filter(l => l.trim() && !l.trim().startsWith('//')).length,
        cyclomaticComplexity,
        complexityRating: cyclomaticComplexity < 10 ? 'low' : cyclomaticComplexity < 20 ? 'medium' : 'high',
        functionCount,
        longFunctions,
        suggestions: [
            ...(cyclomaticComplexity > 20 ? ['Consider breaking down complex logic into smaller functions'] : []),
            ...(longFunctions.length > 0 ? [`${longFunctions.length} function(s) exceed 50 lines - consider refactoring`] : []),
        ],
    };
}

const analyzeComplexity = tool(
  'analyze_complexity',
  'Analyze cyclomatic complexity and other code metrics for files',
  {
    filePath: z.string().describe('Path to the file to analyze'),
  },
  async ({ filePath }) => {
    const result = await _analyzeComplexityImpl(filePath);
    return {
      content: [{
        type: 'text' as const,
        text: JSON.stringify(result, null, 2),
      }],
    };
  }
);

/**
 * Suggest refactorings based on code analysis
 */
export async function _suggestRefactoringImpl(filePath: string, focusArea: string = 'all') {
    const content = await fs.readFile(filePath, 'utf-8');
    const lines = content.split('\n');
    const suggestions: Array<{
      type: string;
      line: number;
      description: string;
      severity: string;
      suggestion: string;
    }> = [];

    // Check for naming issues
    if (focusArea === 'all' || focusArea === 'naming') {
      const singleLetterVars = content.match(/\b(let|const|var)\s+[a-z]\s*=/g) || [];
      if (singleLetterVars.length > 0) {
        suggestions.push({
          type: 'improve-naming',
          line: 0,
          description: `Found ${singleLetterVars.length} single-letter variable names`,
          severity: 'medium',
          suggestion: 'Use descriptive variable names that convey meaning',
        });
      }
    }

    // Check for structural issues
    if (focusArea === 'all' || focusArea === 'structure') {
      // Deeply nested code
      let maxIndent = 0;
      let deepestLine = 0;
      lines.forEach((line, i) => {
        const indent = line.search(/\S/);
        if (indent > maxIndent) {
          maxIndent = indent;
          deepestLine = i + 1;
        }
      });

      if (maxIndent > 20) { // More than 5 levels of nesting (4 spaces each)
        suggestions.push({
          type: 'simplify-conditional',
          line: deepestLine,
          description: `Deep nesting detected (${Math.floor(maxIndent / 4)} levels)`,
          severity: 'high',
          suggestion: 'Consider early returns, guard clauses, or extracting nested logic',
        });
      }

      // Long parameter lists
      const longParams = content.match(/\([^)]{100,}\)/g) || [];
      if (longParams.length > 0) {
        suggestions.push({
          type: 'extract-class',
          line: 0,
          description: 'Functions with many parameters detected',
          severity: 'medium',
          suggestion: 'Consider using an options object or extracting a class',
        });
      }
    }

    // Check for potential dead code
    const todoComments = (content.match(/\/\/\s*(TODO|FIXME|HACK|XXX)/gi) || []).length;
    const commentedCode = (content.match(/\/\/\s*(const|let|var|function|if|for|while)/g) || []).length;

    if (commentedCode > 3) {
      suggestions.push({
        type: 'remove-dead-code',
        line: 0,
        description: `Found ${commentedCode} instances of commented-out code`,
        severity: 'low',
        suggestion: 'Remove commented-out code - use version control for history',
      });
    }

    return {
      file: filePath,
      totalSuggestions: suggestions.length,
      todoComments,
      suggestions,
    };
}

const suggestRefactoring = tool(
  'suggest_refactoring',
  'Analyze code and suggest specific refactoring opportunities',
  {
    filePath: z.string().describe('Path to the file to analyze'),
    focusArea: z.enum(['all', 'naming', 'structure', 'duplication', 'complexity']).optional(),
  },
  async ({ filePath, focusArea = 'all' }) => {
    const result = await _suggestRefactoringImpl(filePath, focusArea);
    return {
      content: [{
        type: 'text' as const,
        text: JSON.stringify(result, null, 2),
      }],
    };
  }
);

/**
 * Create the refactoring tools MCP server
 */
export function createRefactorToolsServer() {
  return createSdkMcpServer({
    name: 'refactor-tools',
    version: '1.0.0',
    tools: [findDuplicates, analyzeComplexity, suggestRefactoring],
  });
}

export { findDuplicates, analyzeComplexity, suggestRefactoring };

File: INTeract-ive-Agent/python/src/tools/refactor_tools.py
"""Refactoring analysis tools for the agent."""

import json
import os
import re
from collections import defaultdict
from pathlib import Path
from typing import Any

from claude_agent_sdk import tool, create_sdk_mcp_server


async def _find_duplicates_impl(directory: str, min_lines: int = 5, file_pattern: str = "*") -> dict[str, Any]:
    """Find duplicate code blocks across files."""
    files: list[Path] = []

    def walk_dir(dir_path: Path) -> None:
        try:
            for entry in dir_path.iterdir():
                if entry.is_dir():
                    if not entry.name.startswith(".") and entry.name != "node_modules":
                        walk_dir(entry)
                elif entry.is_file():
                    if file_pattern == "*" or entry.name.endswith(
                        file_pattern.replace("*", "")
                    ):
                        files.append(entry)
        except PermissionError:
            pass

    walk_dir(Path(directory))

    block_hashes: dict[str, list[dict]] = defaultdict(list)

    for file_path in files:
        try:
            content = file_path.read_text(encoding="utf-8", errors="ignore")
            lines = content.split("\n")

            for i in range(len(lines) - min_lines + 1):
                block = lines[i : i + min_lines]
                block_text = "\n".join(block).strip()

                if len(block_text) < 50:
                    continue

                # Use a stable hash (like hashlib) instead of hash() as it's not stable across runs in Python 3
                import hashlib
                block_hash = hashlib.md5(block_text.encode()).hexdigest()[:16]

                block_hashes[block_hash].append({
                    "file": str(file_path.relative_to(directory)),
                    "start_line": i + 1,
                    "lines": block[:3],
                })
        except Exception:
            pass

    duplicates = [
        {
            "occurrences": len(locations),
            "locations": [
                {
                    "file": loc["file"],
                    "start_line": loc["start_line"],
                    "preview": "\n".join(loc["lines"]),
                }
                for loc in locations
            ],
        }
        for locations in block_hashes.values()
        if len(locations) > 1
    ][:20]

    return {
        "content": [{
            "type": "text",
            "text": json.dumps({
                "files_scanned": len(files),
                "duplicates_found": len(duplicates),
                "duplicates": duplicates,
            }, indent=2),
        }]
    }


@tool(
    "find_duplicates",
    "Find duplicate or similar code blocks across files in a directory",
    {"directory": str, "min_lines": int, "file_pattern": str},
)
async def find_duplicates(args: dict[str, Any]) -> dict[str, Any]:
    return await _find_duplicates_impl(
        args["directory"], 
        args.get("min_lines", 5), 
        args.get("file_pattern", "*")
    )


async def _analyze_complexity_impl(file_path: str) -> dict[str, Any]:
    """Analyze code complexity."""
    content = Path(file_path).read_text(encoding="utf-8", errors="ignore")
    lines = content.split("\n")

    # Simple complexity analysis
    complexity = 1
    complexity_patterns = [
        r"\bif\b",
        r"\belse\b",
        r"\bwhile\b",
        r"\bfor\b",
        r"\bcase\b",
        r"\bcatch\b",
        r"\bexcept\b",
        r"\?",
        r"&&",
        r"\|\|",
        r"\?\?",
        r"\band\b",
        r"\bor\b",
    ]

    for line in lines:
        for pattern in complexity_patterns:
            complexity += len(re.findall(pattern, line))

    # Count functions
    function_patterns = [
        r"def\s+\w+",
        r"async\s+def\s+\w+",
        r"function\s+\w+",
        r"=>\s*\{",
        r"\w+\s*\([^)]*\)\s*\{",
    ]
    function_count = sum(
        len(re.findall(p, content)) for p in function_patterns
    )

    # Find long functions
    long_functions: list[dict] = []
    function_lines = 0
    current_function = None
    func_start_line = 0

    for i, line in enumerate(lines):
        func_match = re.match(r"^(async\s+)?def\s+(\w+)", line.lstrip())
        if func_match:
            if current_function:
                # Close previous function
                length = i - func_start_line
                if length > 50:
                    long_functions.append({"name": current_function, "lines": length})
            
            current_function = func_match.group(2)
            func_start_line = i
        elif current_function and line.strip() and not line.startswith(" ") and not line.startswith("\t"):
            # If line is not indented and not empty, it might be the end of the function
            # This is a simple heuristic for Python
            length = i - func_start_line
            if length > 50:
                long_functions.append({"name": current_function, "lines": length})
            current_function = None

    # Handle the last function in the file
    if current_function:
        length = len(lines) - func_start_line
        if length > 50:
            long_functions.append({"name": current_function, "lines": length})

    rating = "low" if complexity < 10 else "medium" if complexity < 20 else "high"

    suggestions = []
    if complexity > 20:
        suggestions.append("Consider breaking down complex logic into smaller functions")
    if function_count > 20:
        suggestions.append("Many functions detected - consider splitting into modules")
    if long_functions:
        suggestions.append(f"{len(long_functions)} function(s) exceed 50 lines - consider refactoring")

    return {
        "content": [{
            "type": "text",
            "text": json.dumps({
                "file": file_path,
                "total_lines": len(lines),
                "code_lines": len([l for l in lines if l.strip() and not l.strip().startswith("#")]),
                "cyclomatic_complexity": complexity,
                "complexity_rating": rating,
                "function_count": function_count,
                "long_functions": long_functions,
                "suggestions": suggestions,
            }, indent=2),
        }]
    }


@tool(
    "analyze_complexity",
    "Analyze cyclomatic complexity and other code metrics for files",
    {"file_path": str},
)
async def analyze_complexity(args: dict[str, Any]) -> dict[str, Any]:
    return await _analyze_complexity_impl(args["file_path"])


async def _suggest_refactoring_impl(file_path: str, focus_area: str = "all") -> dict[str, Any]:
    """Suggest refactorings for a file."""
    content = Path(file_path).read_text(encoding="utf-8", errors="ignore")
    lines = content.split("\n")
    suggestions: list[dict] = []

    # Check for naming issues
    if focus_area in ("all", "naming"):
        single_letter_vars = re.findall(r"\b(let|const|var|def)\s+[a-z]\s*=", content)
        if single_letter_vars:
            suggestions.append({
                "type": "improve-naming",
                "line": 0,
                "description": f"Found {len(single_letter_vars)} single-letter variable names",
                "severity": "medium",
                "suggestion": "Use descriptive variable names that convey meaning",
            })

    # Check for structural issues
    if focus_area in ("all", "structure"):
        max_indent = 0
        deepest_line = 0
        for i, line in enumerate(lines):
            indent = len(line) - len(line.lstrip())
            if indent > max_indent:
                max_indent = indent
                deepest_line = i + 1

        if max_indent > 20:
            suggestions.append({
                "type": "simplify-conditional",
                "line": deepest_line,
                "description": f"Deep nesting detected ({max_indent // 4} levels)",
                "severity": "high",
                "suggestion": "Consider early returns, guard clauses, or extracting nested logic",
            })

    # Check for dead code
    todo_comments = len(re.findall(r"#\s*(TODO|FIXME|HACK|XXX)", content, re.I))
    commented_code = len(re.findall(r"#\s*(def|class|if|for|while|import)", content))

    return {
        "content": [{
            "type": "text",
            "text": json.dumps({
                "file": file_path,
                "total_suggestions": len(suggestions),
                "todo_comments": todo_comments,
                "suggestions": suggestions,
            }, indent=2),
        }]
    }

@tool(
    "suggest_refactoring",
    "Analyze code and suggest specific refactoring opportunities",
    {"file_path": str, "focus_area": str},
)
async def suggest_refactoring(args: dict[str, Any]) -> dict[str, Any]:
    return await _suggest_refactoring_impl(
        args["file_path"], 
        args.get("focus_area", "all")
    )
    return {
        "content": [{
            "type": "text",
            "text": json.dumps({
                "file": file_path,
                "total_suggestions": len(suggestions),
                "todo_comments": todo_comments,
                "suggestions": suggestions,
            }, indent=2),
        }]
    }


def create_refactor_tools_server():
    """Create the refactoring tools MCP server."""
    return create_sdk_mcp_server(
        name="refactor-tools",
        version="1.0.0",
        tools=[find_duplicates, analyze_complexity, suggest_refactoring],
    )
